File: 5 - Coding Interview Workshops/Week8/Source Code/lesson.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None

    def insert_at_beginning(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def insert_after(self, node, data):
        new_node = Node(data)
        new_node.next = node.next
        node.next = new_node

    def delete_at_beginning(self):
        if self.head is None:
            return
        self.head = self.head.next

    def delete_at_end(self):
        if self.head is None:
            return
        if self.head.next is None:
            self.head = None
            return
        current = self.head
        while current.next.next:
            current = current.next
        current.next = None

    def delete_node(self, node):
        if node is None or node.next is None:
            return
        node.data = node.next.data
        node.next = node.next.next

    def search(self, data):
        current = self.head
        while current:
            if current.data == data:
                return current
            current = current.next
        return None

    def traverse(self):
        current = self.head
        while current:
            print(current.data, end=" ")
            current = current.next

class DoublyLinkedListNode:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert_at_beginning(self, data):
        new_node = DoublyLinkedListNode(data)
        new_node.next = self.head
        if self.head:
            self.head.prev = new_node
        else:
            self.tail = new_node
        self.head = new_node

    def insert_at_end(self, data):
        new_node = DoublyLinkedListNode(data)
        if self.tail is None:
            self.head = new_node
            self.tail = new_node
            return
        new_node.prev = self.tail
        self.tail.next = new_node
        self.tail = new_node

    def insert_after(self, node, data):
        new_node = DoublyLinkedListNode(data)
        new_node.next = node.next
        new_node.prev = node
        if node.next:
            node.next.prev = new_node
        else:
            self.tail = new_node
        node.next = new_node

    def delete_at_beginning(self):
        if self.head is None:
            return
        self.head = self.head.next
        if self.head:
            self.head.prev = None
        else:
            self.tail = None

    def delete_at_end(self):
        if self.tail is None:
            return
        self.tail = self.tail.prev
        if self.tail:
            self.tail.next = None
        else:
            self.head = None

    def delete_node(self, node):
        if node is None:
            return
        if node.prev:
            node.prev.next = node.next
        else:
            self.head = node.next
        if node.next:
            node.next.prev = node.prev
        else:
            self.tail = node.prev

    def search(self, data):
        current = self.head
        while current:
            if current.data == data:
                return current
            current = current.next
        return None

    def traverse(self):
        current = self.head
        while current:
            print(current.data, end=" ")
            current = current.next

def reverse_linked_list(head):
    prev = None
    current = head
    while current is not None:
        next_node = current.next
        current.next = prev
        prev = current
        current = next_node
    return prev

def detect_loop(head):
    slow = head
    fast = head
    while slow and fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

def find_middle(head):
    slow = head
    fast = head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
    return slow

import chardet

def detect_encoding(file_path):
    with open(file_path, 'rb') as file:
        result = chardet.detect(file.read())
        return result['encoding']

def convert_encoding(text, from_encoding, to_encoding):
    return text.encode(from_encoding).decode(to_encoding)

import unicodedata

def normalize_unicode(text, form='NFC'):
    return unicodedata.normalize(form, text)

def case_fold(text):
    return text.casefold()

import re

def replace_unicode_characters(text, property, replacement):
    pattern = re.compile(f'\\p{{{property}}}')
    return pattern.sub(replacement, text)

File: 5 - Coding Interview Workshops/Week08/Source Code/lesson.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
  }

  insert_at_beginning(data) {
    const new_node = new Node(data);
    new_node.next = this.head;
    this.head = new_node;
  }

  insert_at_end(data) {
    const new_node = new Node(data);
    if (this.head === null) {
      this.head = new_node;
      return;
    }
    let current = this.head;
    while (current.next) {
      current = current.next;
    }
    current.next = new_node;
  }

  insert_after(node, data) {
    const new_node = new Node(data);
    new_node.next = node.next;
    node.next = new_node;
  }

  delete_at_beginning() {
    if (this.head === null) {
      return;
    }
    this.head = this.head.next;
  }

  delete_at_end() {
    if (this.head === null) {
      return;
    }
    if (this.head.next === null) {
      this.head = null;
      return;
    }
    let current = this.head;
    while (current.next.next) {
      current = current.next;
    }
    current.next = null;
  }

  delete_node(node) {
    if (node === null || node.next === null) {
      return;
    }
    node.data = node.next.data;
    node.next = node.next.next;
  }

  search(data) {
    let current = this.head;
    while (current) {
      if (current.data === data) {
        return current;
      }
      current = current.next;
    }
    return null;
  }

  traverse() {
    let current = this.head;
    while (current) {
      console.log(current.data);
      current = current.next;
    }
  }
}

class DoublyLinkedListNode {
  constructor(data) {
    this.data = data;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  insert_at_beginning(data) {
    const new_node = new DoublyLinkedListNode(data);
    new_node.next = this.head;
    if (this.head) {
      this.head.prev = new_node;
    } else {
      this.tail = new_node;
    }
    this.head = new_node;
  }

  insert_at_end(data) {
    const new_node = new DoublyLinkedListNode(data);
    if (this.tail === null) {
      this.head = new_node;
      this.tail = new_node;
      return;
    }
    new_node.prev = this.tail;
    this.tail.next = new_node;
    this.tail = new_node;
  }

  insert_after(node, data) {
    const new_node = new DoublyLinkedListNode(data);
    new_node.next = node.next;
    new_node.prev = node;
    if (node.next) {
      node.next.prev = new_node;
    } else {
      this.tail = new_node;
    }
    node.next = new_node;
  }

  delete_at_beginning() {
    if (this.head === null) {
      return;
    }
    this.head = this.head.next;
    if (this.head) {
      this.head.prev = null;
    } else {
      this.tail = null;
    }
  }

  delete_at_end() {
    if (this.tail === null) {
      return;
    }
    this.tail = this.tail.prev;
    if (this.tail) {
      this.tail.next = null;
    } else {
      this.head = null;
    }
  }

  delete_node(node) {
    if (node === null) {
      return;
    }
    if (node.prev) {
      node.prev.next = node.next;
    } else {
      this.head = node.next;
    }
    if (node.next) {
      node.next.prev = node.prev;
    } else {
      this.tail = node.prev;
    }
  }

  search(data) {
    let current = this.head;
    while (current) {
      if (current.data === data) {
        return current;
      }
      current = current.next;
    }
    return null;
  }

  traverse() {
    let current = this.head;
    while (current) {
      console.log(current.data);
      current = current.next;
    }
  }
}

function reverseLinkedList(head) {
  let prev = null;
  let current = head;
  while (current !== null) {
    const next_node = current.next;
    current.next = prev;
    prev = current;
    current = next_node;
  }
  return prev;
}

function detectLoop(head) {
  let slow = head;
  let fast = head;
  while (slow && fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
    if (slow === fast) {
      return true;
    }
  }
  return false;
}

function findMiddle(head) {
  let slow = head;
  let fast = head;
  while (fast !== null && fast.next !== null) {
    slow = slow.next;
    fast = fast.next.next;
  }
  return slow;
}

const chardet = require("chardet");
const fs = require("fs");

function detectEncoding(filePath) {
  const buffer = fs.readFileSync(filePath);
  return chardet.detect(buffer);
}

function convertEncoding(text, fromEncoding, toEncoding) {
  const buffer = Buffer.from(text, fromEncoding);
  return buffer.toString(toEncoding);
}

function normalizeUnicode(text, form = "NFC") {
  return text.normalize(form);
}

function caseFold(text) {
  return text.toLowerCase();
}

function replaceUnicodeCharacters(text, property, replacement) {
  const pattern = new RegExp(`\\p{${property}}`, "gu");
  return text.replace(pattern, replacement);
}
